lib/spack/spack/build_systems/cached_cmake.py

```diff
@@ -195,26 +231,57 @@ def initconfig_hardware_entries(self):
             "#------------------{0}\n".format("-" * 60),
         ]
 
+        # Provide standard CMake arguments for dependent CachedCMakePackages
         if spec.satisfies("^cuda"):
             entries.append("#------------------{0}".format("-" * 30))
             entries.append("# Cuda")
             entries.append("#------------------{0}\n".format("-" * 30))
 
             cudatoolkitdir = spec["cuda"].prefix
-            entries.append(cmake_cache_path("CUDA_TOOLKIT_ROOT_DIR", cudatoolkitdir))
-            cudacompiler = "${CUDA_TOOLKIT_ROOT_DIR}/bin/nvcc"
-            entries.append(cmake_cache_path("CMAKE_CUDA_COMPILER", cudacompiler))
+            entries.append(cmake_cache_path("CUDAToolkit_ROOT", cudatoolkitdir))
+            entries.append(cmake_cache_path("CMAKE_CUDA_COMPILER", "${CUDAToolkit_ROOT}/bin/nvcc"))
             entries.append(cmake_cache_path("CMAKE_CUDA_HOST_COMPILER", "${CMAKE_CXX_COMPILER}"))
 
+            archs = spec.variants["cuda_arch"].value
+            if archs != "none":
+                arch_str = ";".join(archs)
+                entries.append(
+                    cmake_cache_string("CMAKE_CUDA_ARCHITECTURES", "{0}".format(arch_str))
+                )
+
+        # Since there is no rocm package, look for the variant
+        if "+rocm" in spec:
+            entries.append("#------------------{0}".format("-" * 30))
+            entries.append("# ROCm")
+            entries.append("#------------------{0}\n".format("-" * 30))
+
+            # Explicitly setting HIP_ROOT_DIR may be a patch that is no longer necessary
+            entries.append(cmake_cache_path("HIP_ROOT_DIR", "{0}".format(spec["hip"].prefix)))
+            entries.append(
+                cmake_cache_path("HIP_CXX_COMPILER", "{0}".format(self.spec["hip"].hipcc))
+            )
+            archs = self.spec.variants["amdgpu_target"].value
+            if archs != "none":
+                arch_str = ";".join(archs)
+                entries.append(
+                    cmake_cache_string("CMAKE_HIP_ARCHITECTURES", "{0}".format(arch_str))
+                )
+                entries.append(cmake_cache_string("AMDGPU_TARGETS", "{0}".format(arch_str)))
+                entries.append(cmake_cache_string("GPU_TARGETS", "{0}".format(arch_str)))
+
         return entries
 
     def std_initconfig_entries(self):
+        cmake_prefix_path_env = os.environ["CMAKE_PREFIX_PATH"]
+        cmake_prefix_path_array = cmake_prefix_path_env.split(":")
+        cmake_prefix_path = ";".join(cmake_prefix_path_array)
         return [
             "#------------------{0}".format("-" * 60),
             "# !!!! This is a generated file, edit at own risk !!!!",
             "#------------------{0}".format("-" * 60),
             "# CMake executable path: {0}".format(self.pkg.spec["cmake"].command.path),
             "#------------------{0}\n".format("-" * 60),
+            cmake_cache_path("CMAKE_PREFIX_PATH", cmake_prefix_path),
```

Comment:
Note that `CachedCMakePackage` already sets `CMAKE_PREFIX_PATH` as a `cmake` command line option (the behavior is inherited from `CMakePackage`): did you also need to set it in the cache?

Reply:
@scheibelp While it may set it as a command line option, if you run cmake outside of the spack environment that created it, it will not be able to find all of the spack created dependencies.  Preserving the prefix path inside of the `.cmake` file allows you to reuse the file afterwards and without spack being actively involved.

Reply:
OK: I assume the same principle applies to the generator (which is handled in `cmake.py` as a command-line option, but specifying it in the settings file avoids the need to set it when invoking `cmake` outside of Spack.

Reply:
@scheibelp That is correct. It will be cached in the `.cmake` file so that when you invoke it you have a command that looks like 
```
cmake -C<filename>.cmake -B <path>/build -DCMAKE_INSTALL_PREFIX=<path>/install .
```

Reply:
This is actually not a path. [It's most appropriately a string in CMake](https://cmake.org/cmake/help/git-stage/command/set.html#set-cache-entry). A `PATH` is a path to a directory (note "a directory", not a list of directories).